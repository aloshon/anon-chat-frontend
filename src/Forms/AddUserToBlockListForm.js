import React, { useContext } from "react";
import useFields from '../Hooks/useFields';
import "./Forms.css";
import AnonChatApi from "../api";
import Container from 'react-bootstrap/Container';
import Button from 'react-bootstrap/Button';
import Form from 'react-bootstrap/Form';
import {useDispatch} from 'react-redux';
import UserContext from "../UserContext";
import {blockUser} from "../Actions/actionCreators";

/**
 * AddUserToBlockListForm component renders form for blocking users
 * by username
 * If user is already blocked, send an alert
 */
const AddUserToBlockListForm = ({blockList}) => {
    const dispatch = useDispatch();
    const {user} = useContext(UserContext)
    const [formData, handleChange, resetFormData] = useFields({
        username: ''
    });

    const handleSubmit = async (e) => {
        e.preventDefault();
        try{
            // Make sure user is not blocking themselves, someone twice, 
            // or if the other user blocked them
            if(formData.username === user.username) return

            let userToBlock = await AnonChatApi.checkForUser(formData.username);
            if(!userToBlock){
                throw `No user with username: ${formData.username}`
            }

            let alreadyBlocked = blockList.find(u => u.blocked_username === formData.username);
            if(alreadyBlocked) throw `User already blocked ${formData.username}`;

            dispatch(blockUser(formData.username));
            resetFormData();
            return
        } catch(e){
            console.error(`Error blocking user! ${e}`)
            alert(`Error blocking user! ${e}`);
        }
    }
    return (
        <Container>
            <h3 style={{ textAlign: "left", margin: "15px"}}>Block User</h3>
            <Form onSubmit={handleSubmit}>
                <Form.Group>
                    <Form.Control name="username" type="text" value={formData.username} onChange={handleChange} placeholder="Username" />
                </Form.Group>
                <Button variant="primary" type="submit" size="md" block>
                    Submit
                </Button>
            </Form>
        </Container>
    )
}

export default AddUserToBlockListForm;