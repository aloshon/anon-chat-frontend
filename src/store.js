import  { composeWithDevTools } from "redux-devtools-extension";
import { persistStore, persistReducer } from "redux-persist";
import storage from "redux-persist/lib/storage";
import autoMergeLevel2 from "redux-persist/lib/stateReconciler/autoMergeLevel2";

import thunk from "redux-thunk";
import rootReducer from "./Reducers/rootReducer";
import { createStore, applyMiddleware } from "redux";

// only keep the darkMode reducer persistent in local storage
const persistConfig = {
  key: "root",
  whitelist:['darkMode'],
  storage,
  stateReconciler: autoMergeLevel2
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = createStore(
  persistedReducer,
  composeWithDevTools(
    applyMiddleware(thunk),
  )
);

export const persistedStore = persistStore(store);