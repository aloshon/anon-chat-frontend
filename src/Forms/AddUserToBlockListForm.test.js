import { render, screen } from '@testing-library/react';
import AddUserToBlockListForm from './AddUserToBlockListForm';
import { Provider } from "react-redux";
import { store } from "../store"
import {BrowserRouter} from 'react-router-dom';
import UserContext from "../UserContext";


describe("AddUserToBlockListForm should render form to block user", () => {
    const user = {
        id: 1,
        username: "user1",
        isAdmin: false,
        blockList: []
    }

    const {asFragment} = render(
        <Provider store={store}>
            <BrowserRouter>
                <UserContext.Provider value={{user}}>
                    <AddUserToBlockListForm />
                </UserContext.Provider>
            </BrowserRouter>
      </Provider>
    );
    test("renders inputs", () => {
      const placeholder1 = screen.getByPlaceholderText("Username");
      const button = screen.getByText("Submit");
      expect(placeholder1).toBeInTheDocument();
      expect(button).toBeInTheDocument();
    });
  
    test("matches snapshot", () => {
      expect(asFragment()).toMatchSnapshot();
    });
    
  });