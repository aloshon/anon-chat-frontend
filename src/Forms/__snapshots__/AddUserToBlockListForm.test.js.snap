// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddUserToBlockListForm should render form to block user matches snapshot 1`] = `<DocumentFragment />`;
